text/0000-project-toolchain-shims.md

```diff
@@ -0,0 +1,69 @@
+
+This same scan will also occur when pinning a version of Node using `notion use`, to ensure that shims are created for packages installed prior to setting up a toolchain for the project.  Finally, the scan will be available via a new `notion autoshim .` command.
+
+Once the scan has completed, Notion will exit with a successful status, even if shims were unable to be created, as per [RFC 0020](0020-exit-codes.md).
+
+Due to potential future changes (see [Critique](#critique) below), it is important that scanning be implemented in a modular fashion, such that a full scan, single dependency scan, or single shim creation can be kicked off independently.
+
+# Critique
+[critique]: #critique
+
+## "Full scan" vs. "smart scan"
+
+This RFC proposes that all shims be checked every time `yarn` or `npm` is run, but alternatives exist.  In particular, the amount of scanning being done could be significantly reduced by parsing the Yarn/npm command line arguments and only checking the shims actually being installed during `yarn add <module(s)>` or `npm install <module(s)>`.
+
+However, it will always be necessary to implement a "full scan" scenario for the initial install of modules — i.e. `yarn install` and `npm install`.  And while a "smart scan" would increase speed, it is also potentially brittle in multiple ways.  It would require Notion to both effectively duplicate those tools' command line parsers and maintain an understanding of their command structures.  It would also fail to fill "holes" in Notion's shim coverage — once a shim has been missed, by any means, it will not be created until a full scan is done.
+
+## Sync vs. async
+
+Another alternative would be to run the scanning and shim creation process in the background, allowing an earlier return from the `yarn` and `npm` shims.  Asynchrony here would make Notion feel more responsive, but introduce a race condition where the shims are not available immediately after the command completes.  The two deciding factors in favor of synchronous shim creation are:
+
+1. If shim creation is fast enough to complete between when the `yarn`/`npm` shim completes and when the user immediately thereafter tries to use one of the created shims, it *should* also be fast enough to run synchronously without affecting the "speed feel".
+2. A race condition introduces the possibility for command chaining to fail:
+
+```
+$ yarn install && mocha
+-bash: mocha: command not found
+$ mocha
+  example
+    ✓ runs a test
+
+
+  1 passing (11ms)
+```
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- When, if ever, should shims be deleted?  The ability of a user to simply delete an entire project from disk makes tracking shim use problematic, at best.  What, exactly, is the behavior of a shim when the associated tool is not part of the toolchain?  If it simply "forwards" to the shell, this is less of an issue.
```

Comment:
My first instinct is to say this isn't a problem, since shims are more or less transparent. Yes, typing `which foo` will show a different answer than if Notion weren't installed, but otherwise the shim is supposed to behave identically to the underlying tool it delegates to (and `notion which foo` is a way to "see through" that difference as well), and be fast at doing its delegation. So generally I think it should be fine to leave shims in place.

The counterargument would be that shims are the most intrusive part of the Notion design, in that they impact the way developers think about how tools work on their machine.

But the problem with that is, if someone wants to automatically register a shim and projects keep automatically re-creating the shim, they're going to be really frustrated.

So maybe a coarser-grained approach is the right answer: if Notion is operational, you are letting it control the set of shims. If you want to get the shims out of your way, you should deactivate Notion all at once, with `notion deactivate`.

Reply:
> If it simply "forwards" to the shell, this is less of an issue.

Yes, the behavior should be forwarding to the shell. So the shims are effectively fully transparent except for the behavior of `which` and the process tree (for example when inspecting the process table).